### PR TITLE
Fix broadcasting by using vector SkyCoords internally [WIP]

### PR DESCRIPTION
**Pull request by StuartLittlefair**
As described in #215, I think there's an argument for all internal calculations converting their arguments to ```Time``` and ```SkyCoord``` objects, which may be vector. This PR builds on the utility function in #215 and implements this functionality. It's a large change in the ```observer``` codebase, but outward facing functionality should be compatible with how it was before. However, there are so many unassociated test failures from other issues with astropy 1.3 it's hard to know, hence the WIP designation.

Once complete, this PR *could* fix remaining broadcasting issues with astropy 1.3 (e.g #282, #263, #272).

Some notes:

- This PR *requires* astropy 1.3 as a minimum version.
- ~~Although behaviour for scalar times and targets is unchanged, there is an improvement (IMO) in functionality for multiple times and targets. I briefly describe that here.~~

**This PR should give identical behaviour to the existing astroplan code, though the internals of the ```observer``` module have changed quite a lot. Please disregard the code samples below - I've left them in the comment for historical purposes**

First - for scalar objects:
```python
In [2]:  apo = Observer.at_site("APO")
In [3]:  times = Time(2457786.9048868627, format='jd')
In [4]:  m31 = coord.SkyCoord.from_name('M31')
In [5]:  apo.altaz(times, m31)
Out[5]: 
<SkyCoord (AltAz: obstime=2017-02-02 09:43:02.224949, location=(-1463969.3018517173, -5166673.342234327, 3434985.7120456537) m, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0, obswl=1.0 micron): (az, alt) in deg
    ( 346.02166236, -13.98494769)>
```

However, if the times and targets are 1D and broadcastable, altaz and similar calcs will broadcast automatically:
```python
In [6]: ippeg = coord.SkyCoord.from_name('IP Peg')
In [7]: deneb = coord.SkyCoord.from_name('Deneb')
In [8]: times = times + np.linspace(-0.3, 0.3, 3)*u.d
In [9]: apo.altaz(times, [m31, ippeg, deneb])
Out[9]: 
<SkyCoord (AltAz: obstime=[datetime.datetime(2017, 2, 2, 2, 31, 2, 224949)
 datetime.datetime(2017, 2, 2, 9, 43, 2, 224949)
 datetime.datetime(2017, 2, 2, 16, 55, 2, 224949)], location=(-1463969.3018517173, -5166673.342234327, 3434985.7120456537) m, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0, obswl=1.0 micron): (az, alt) in deg
    [( 297.48755926,  47.14479827), (   2.09385352, -38.68372101),
     (  52.36092567,  64.09699143)]>
```

also works with lists of ```FixedTargets```

```python
In [11]: apo.altaz(times, [FixedTarget(targ) for targ in [m31, ippeg, deneb]])
Out[11]: 
<SkyCoord (AltAz: obstime=[datetime.datetime(2017, 2, 2, 2, 31, 2, 224949)
 datetime.datetime(2017, 2, 2, 9, 43, 2, 224949)
 datetime.datetime(2017, 2, 2, 16, 55, 2, 224949)], location=(-1463969.3018517173, -5166673.342234327, 3434985.7120456537) m, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0, obswl=1.0 micron): (az, alt) in deg
    [( 297.48755926,  47.14479827), (   2.09385352, -38.68372101),
     (  52.36092567,  64.09699143)]>
```

So far the same. But using vector skycoords internally adds functionality, since the ```horizon``` argument to the ```observer``` routines can be non-scalar and it will work as long as you can broadcast the horizon and altaz results:

```python
In [12]: apo.target_is_up(times, [m31, ippeg, deneb], horizon=u.Quantity([50*u.deg, 0*u.deg, 20*u.deg]))
Out[12]: array([False, False,  True], dtype=bool)
In [13]: apo.target_is_up(times, [m31, ippeg, deneb])
Out[13]: array([ True, False,  True], dtype=bool)
```

Also, by manipulating the shape of the times and targets object you can create a grid of ```altaz``` at all times and targets, rather than broadcasting, e.g:

```python
In [19]: apo.altaz(times[:, np.newaxis], [m31, ippeg, deneb])
Out[19]: 
<SkyCoord (AltAz: obstime=[[datetime.datetime(2017, 2, 2, 2, 31, 2, 224949)
  datetime.datetime(2017, 2, 2, 2, 31, 2, 224949)
  datetime.datetime(2017, 2, 2, 2, 31, 2, 224949)]
 [datetime.datetime(2017, 2, 2, 9, 43, 2, 224949)
  datetime.datetime(2017, 2, 2, 9, 43, 2, 224949)
  datetime.datetime(2017, 2, 2, 9, 43, 2, 224949)]
 [datetime.datetime(2017, 2, 2, 16, 55, 2, 224949)
  datetime.datetime(2017, 2, 2, 16, 55, 2, 224949)
  datetime.datetime(2017, 2, 2, 16, 55, 2, 224949)]], location=(-1463969.3018517173, -5166673.342234327, 3434985.7120456537) m, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0, obswl=1.0 micron): (az, alt) in deg
    [[( 297.48755926,  47.14479827), ( 277.5507293 ,  23.53321076),
      ( 319.5051457 ,   8.36398593)],
     [( 346.02166236, -13.98494769), (   2.09385352, -38.68372101),
      (  28.22799217,  -3.00973326)],
     [(  53.52265937,  21.03003458), (  84.15855299,  26.41118594),
      (  52.36092567,  64.09699143)]]>
```

Although I haven't timed the performance I suspect it will be faster than the old, non-vectorised code as well.

There are still some issues with using vector ```Times``` or ```SkyCoords``` with higher dimensionality - see TODO note in ```_horiz_cross```.

### CODE REVIEW

**Comment by StuartLittlefair** [comment]
@bsipocz Can you take a look at this PR? I've so many other issues with my local build (the remote test issues etc) that I can't test it properly, but it may address the broadcasting issues. Is there an easy way to make a local copy of a branch with all the prior fixes so I can test thoroughly? I'm a bit git-challenged!

@bmorris3 - this is a really big change to the code, so you probably want to take a look as well.

**Comment by bsipocz** [comment]
@StuartLittlefair - Sure, I've merged this into the testing branch, test are running here: https://travis-ci.org/bsipocz/astroplan/builds/197593459
There were quite a few conflicts, but hopefully I got them resolved properly.

**Comment by bsipocz** [comment]
The docstring issues seems real, what do you think? https://travis-ci.org/bsipocz/astroplan/jobs/197593463#L759

**Comment by bsipocz** [comment]
Ohh, and I forgot to answer the question: usually for testing multiple PRs, I usually open a new separate test branch and merge in the PRs there. If you have travis turned on your fork, this would do the tests there.

**Comment by StuartLittlefair** [comment]
I think the failures are real. 

There's something that needs discussion here - in previous versions of astroplan if you called altaz with M targets and N times (M /= N), you would get an (M, N) array back. Also, if you called altaz with a list of times and single target you'd get a (1, N) array.

My PR has changed that behaviour, which has caused some failures in the constraints and scheduling code. I *could* re-instate it so that altaz worked as before but I would argue the way it works in this PR is an improvement, and it's better the fix the scheduling and constraints code to work with the new behaviour. 

This does cause the issue that some existing user code might break, but there is still a healthy API warning on the astroplan main page...


This is not the current default behaviour

File: astroplan/tests/test_target.py [hunk 1]
```diff
@@ -40,3 +43,38 @@ def test_FixedTarget_ra_dec():
...
+    subaru = Observer.at_site('subaru')
+    time = Time("2016-01-22 12:00")
+    pos, vel = subaru.location.get_gcrs_posvel(time)
+    gcrs_frame = GCRS(obstime=Time("2016-01-22 12:00"), obsgeoloc=pos, obsgeovel=vel)
+    m31_gcrs = m31.transform_to(gcrs_frame)
+    m31_gcrs_with_distance = m31_with_distance.transform_to(gcrs_frame)
+
+    coo = get_skycoord(m31)
+    assert coo.is_equivalent_frame(ICRS())
+    assert_raises(TypeError, len, coo)
```
Comment (bsipocz) [inline astroplan/tests/test_target.py]: Could you please use pytest.raises instead? We don't have nose as a mandatory requirement but pytest, and this currently fails with ``ImportError: Need nose >= 1.0.0 for tests - see http://somethingaboutorange.com/mrl/projects/nose``

**Comment by bsipocz** [comment]
OK, I see. For that discussion I think we need @bmorris3  and @eteq, as I'm not aware of the details for any decision behind the original API.

**Comment by StuartLittlefair** [comment]
@bsipocz given the currently broken state of astroplan with astropy 1.3, what do you think about making this PR keep the original behaviour, and submit another PR later for the API change?

**Comment by bsipocz** [comment]
Sounds like a good approach to me.

**Comment by StuartLittlefair** [comment]
OK - so the version pushed now restores the original behaviour of astroplan, at least as far as it comes to passing the tests. There may be some cases I've not reverted properly - e.g ```sun_rise_time``` called with N times returns an NxN grid. 

A local test branch with #281, #279, #273 and this PR applied seem to pass tests, with the exception of one failure in scheduler which is (I think) unrelated.

**Comment by bsipocz** [comment]
Testing this with travis now here: https://travis-ci.org/bsipocz/astroplan/builds/197668370

**Comment by bsipocz** [comment]
There are still failures, one is the usage of nose (see inline comment), the other one is unrelated and I can fix in another PR.

**Comment by StuartLittlefair** [comment]
thanks @bsipocz - fixed the nose issue.

**Comment by bsipocz** [comment]
@StuartLittlefair - Thanks a lot, quite good to see that everything is back to normal.

pinging @olebole and @ViviCoder as they may want to use this to patch the debian version

**Comment by bmorris3** [comment]
@StuartLittlefair Unfortunately there's a little rebase that needs to occur after an update to some `Observer` docstrings. Still seem to be some build failures, too.

**Comment by StuartLittlefair** [comment]
@bmorris3 I've just rebased. I don't think this has any chance of passing until #281, #279, #273 are also merged...

**Comment by StuartLittlefair** [comment]
@bmorris3 It might be nice to make astroplan work with the latest version of astropy. I've got two PRs which do that in principle, this and #286. Both only work with astropy v1.3, but #286 could break some user-facing code.

#286 is nearly passing tests - just some docstring issues I can't work out.

OTOH, this PR doesn't include the PRs mentioned in the thread above, so will never pass tests. Those PRs also can't be merged, since they don't pass tests! I'd appreciate some advice on the next step. Is it a good idea to merge in the changes in #281, #279, #273 and get this PR passing?

**Comment by bsipocz** [comment]
@StuartLittlefair @bmorris3 - I think we are already at a point when we need to merge all the bugfix PRs without any of the individual ones passing all the tests. There were already a few tests branches that included all those PRs, if you like I can rerun the tests on the latest one to see whether there are any new divergences compared to astropy dev.

**Comment by StuartLittlefair** [comment]
@bsipocz - I think that would be really helpful.

**Comment by bsipocz** [comment]
OK, so I've rebased the test branch, and merge this on in, tests are here:
https://travis-ci.org/bsipocz/astroplan/builds/200202721

We have the a ``TargetAlwaysUpWarning``, and a real failure with numpy dev, and a few sphinx warnings left.

**Comment by StuartLittlefair** [comment]
@bsipocz so I've worked out how to make a test branch on my own repository :)

With the latest commits I think I've got all the failures except some of the sphinx warnings. In particular I'm not sure what to do about the ```OldEarthOrientationWarning``` . I'm also not sure if the latex-related warnings in the sphinx build are important, but they are part of astropy helpers.

Can you merge my latest commits to my test branch and see if you agree with the summary above. Also, any idea what to do about the ```OldEarthOrientationWarning```s?

**Comment by bsipocz** [comment]
@StuartLittlefair - Latest commits from here or from #286? We had the ``OldEarth...`` before, but only had some workarounds for it never a proper solution. 
There is an astropy-helpers update PR, too, the latex warnings should go away with that.

**Comment by StuartLittlefair** [comment]
@bsipocz from here.

I'm tracking fixes in this PR over there. My hope is that this (or a branch with these commits) can be merged, and I can rebase #286 so the changes are clearer.

**Comment by bsipocz** [comment]
Done. Tests are here: https://travis-ci.org/bsipocz/astroplan/builds/200726025

**Comment by StuartLittlefair** [comment]
@bsipocz do my eyes deceive me or is this a passing branch? 

What's the next step? Presumably one PR based on all the commits that make up your test branch?

**Comment by bsipocz** [comment]
Wow, I haven't noticed, copied the link here while it was still running.
Actually I would prefer the individual PRs being merged as a big joint PR will mess up any future bug hunts, etc. 
Next step is to get the attention of someone with commit rights, and if they want a big joint PR, I can open it straight away from that passing branch ;)

**Comment by StuartLittlefair** [comment]
@bmorris3 @eteq 

Thanks to the hard work of @bsipocz we now have a collection of PRs (this one included) which in combination give a passing test suite! 

Note that contrary to what is said in the initial message with this PR, it does not change the API.

The question is whether you would like multiple small PRs to be merged, or to review on huge PR. In the latter case @bsipocz can create that from her passing branch (see test results above).

**Comment by bsipocz** [comment]
@StuartLittlefair - Could you please rebase this?

**Comment by eteq** [comment]
@StuartLittlefair - I merged all the others, but I think this now needs a rebase?

**Comment by StuartLittlefair** [comment]
@eteq @bsipocz Ok - this has been rebased.

**Comment by eteq** [comment]
I just tried this locally and now that #273 is in it gets all the tests passing with astropy 1.3.

However, the downside is that this now forces 1.3 as a requirement... At a minimum, this should be reflected in the ``setup.py`` (i.e., update ``install_requires`` to be ``'astropy>=1.3'`` instead of 1.2).  But the bigger question is whether it's OK to require 1.3?  

Personally I can see that it would be pretty painful to add a backwards compatibility layer at this point, but I imagine it's possible if absolutely necessary.  But I think it might lead to further confusion down the line to do that, so if a break is necessary it's probably better now... But you'll definitely want to note this in the changelog.

 @bmorris3 - do you have any opinions on this?

**Comment by bmorris3** [comment]
I vote to march forward with 1.3. Viva la revolucion.

**Comment by eteq** [comment]
Alright, cool.  Then @StuartLittlefair, can you update setup.py and add a note to the changelog? With that I think this is good to go!

**Comment by olebole** [comment]
@vivicoder could you check if we can create a bugfix for the Debian package from this so that we have a fully working version there?

**Comment by StuartLittlefair** [comment]
@eteq @bsipocz one issue remaining building the docs which seems to be that astropy=1.3 is not available for numpy 1.12 on conda. is it ok to change travis.yml to use numpy 1.11 for that build or should this be fixed elsewhere?

**Comment by bsipocz** [comment]
That should be working and use pip install this case when an astropy
version is not available. I'll look at it after  I get home tonight

**Comment by bsipocz** [comment]
OK, so the problem there is that there is no wcsaxes build for np 1.12. 
For the sake of this PR, yes please change ``NUMPY_VERSION`` for the docs build to 1.11.

In the meantime I'll open another PR that changes to use wcsaxes from astropy when possible rather than the standalone version. 
But as wcsaxes is already been merged into astropy and s astroplan already requires astropy 1.3, I'm opening a new PR that does the switch.

**Comment by bsipocz** [comment]
At the end wcsaxes isn't directly used so it made sense to open a PR against this one.

**Comment by StuartLittlefair** [comment]
Thanks for the PR @bsipocz In the end it hasn't helped - we now have a similar problem with  astro query.

**Comment by bsipocz** [comment]
@StuartLittlefair - this is indeed an annoying issue. To have a quick and dirty fix, I suggest to add ``conda-forge`` to the ``CONDA_CHANNELS`` variable. For the long term we will come up with something in ``ci-helpers``

**Comment by StuartLittlefair** [comment]
@bmorris3 @eteq @bsipocz OK - we have a passing PR!

**Comment by bmorris3** [comment]
@StuartLittlefair Thanks for all of your hard work on this one!

**Pull request merged**